packages/kokkos/core/unit_test/TestDeepCopy.hpp

```diff
@@ -0,0 +1,167 @@
+#include<Kokkos_Core.hpp>
+
+namespace Test {
+
+namespace Impl {
+template<class MemorySpaceA, class MemorySpaceB>
+struct TestDeepCopy {
+
```

Comment:
Can we add test for non eti'ed data types (user data type)  to make sure if works for such types ?

Reply:
I can't really comment on the Cuda implications here... The original spirit of the patch was to resolve only the hostspace deep_copy issues.

The deep_copy impl I presented is type invariant (which works fine on the host). It has the same signature as std::memcpy, but does parallel execution inside.

I don't think this problem you are having is related to this patch though. It seems to fail with Trilinos/develop without this patch.

Reply:
I agree with you. Ifpack2 failure is not related to this PR nor deep copy on host. I just think if we include non built-in type deep copy, it would be good for testing.  Just in case that byte wise parallel copying somehow copies partial data of a struct from one thread and the other thread copies the other half.